zebra-network/src/config.rs

```diff
@@ -176,3 +184,86 @@ impl Default for Config {
         }
     }
 }
+
+impl<'de> Deserialize<'de> for Config {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields, default)]
+        struct DConfig {
+            listen_addr: String,
+            network: Network,
+            initial_mainnet_peers: HashSet<String>,
+            initial_testnet_peers: HashSet<String>,
+            peerset_initial_target_size: usize,
+            #[serde(alias = "new_peer_interval")]
+            crawl_new_peer_interval: Duration,
+        }
+
+        impl Default for DConfig {
+            fn default() -> Self {
+                let config = Config::default();
+                Self {
+                    listen_addr: config.listen_addr.to_string(),
+                    network: config.network,
+                    initial_mainnet_peers: config.initial_mainnet_peers,
+                    initial_testnet_peers: config.initial_testnet_peers,
+                    peerset_initial_target_size: config.peerset_initial_target_size,
+                    crawl_new_peer_interval: config.crawl_new_peer_interval,
+                }
+            }
+        }
+
+        let config = DConfig::deserialize(deserializer)?;
+        let listen_addr = match config.listen_addr.parse::<SocketAddr>() {
+            Ok(socket) => Ok(socket),
+            Err(_) => match config.listen_addr.parse::<IpAddr>() {
+                Ok(ip) => {
+                    let port = match config.network {
+                        Network::Mainnet => 8233,
+                        Network::Testnet => 18233,
+                    };
+                    Ok(SocketAddr::new(ip, port))
```

Comment:
There's already a default port method :-)
```suggestion
                    Ok(SocketAddr::new(ip, config.network.default_port()))
```

Reply:
Nice! I think I had seen the same match in other places, maybe I need to update them too.

Reply:
Yeah anywhere we match on network just to get ports or strings can be refactored - but please do it in a new PR.

Reply:
I think I need to fix only this: https://github.com/ZcashFoundation/zebra/blob/3f45735f3fda751de0b736bde02e3b330b052a73/zebra-network/src/peer_set/initialize.rs#L129-L132
It's okay to do this in this PR?

Reply:
Sure, and you can make it even simpler:
```rust
    let wrong_net = match config.network {
        Network::Mainnet => Network::Testnet,
        Network::Testnet => Network::Mainnet,
    };
    let wrong_net_port = wrong_net.default.port();
```

Reply:
Yeah, already made it 🙂 did not edit my previous comment.

Reply:
This is really minor, but I just realised that we could make this code a lot more readable with:
```rust
    use Network::*;
    let wrong_net = match config.network {
        Mainnet => Testnet,
        Testnet => Mainnet,
    };
    let wrong_net_port = wrong_net.default_port();
```

Then we're focused on the variant swap, not the enum they come from.